apps/passport-server/src/services/devconnectPretixSyncService.ts

```diff
@@ -46,126 +48,259 @@ interface EventData {
+ * which can be passed to the Pretix fetch functions, in place of the
+ * native 'fetch' API call. Behind the scenes, this function will
+ * increment a counter. When the counter exceeds the limit, the call to
+ * the fetcher function will suspend until the counter is reset. When
+ * this happens, an event is raised. In this system, we use this event
+ * to mark a sync run as "rate-limited", allowing us to complete the run
+ * and resume fetching the remaining data on the next run, when we will
+ * reset the counter.
+ */
+class LimitedFetcher extends EventEmitter {
+  // The maximum number of requests to allow
+  private limit: number;
+  // The number of requests made so far
+  private count: number;
+
+  public constructor(limit: number) {
+    super();
+    this.limit = limit;
+    this.count = 0;
+  }
+
+  public reset(): void {
+    this.count = 0;
+    this.emit("reset");
+  }
 
+  // Returns a function which matches the signature of the 'fetch' API
+  public fetcher(): FetchFn {
+    return async (input, init?) => {
+      if (this.count >= this.limit) {
+        // We are stuck at the limit, so do nothing until reset
+        // By awaiting here, we will pause indefinitely, relying on
+        // another system to reset our counter. In the sync context, the
+        // counter will be reset on each run, allowing us to pick up the
+        // request we were limited from making on the previous run.
+        this.emit("pause");
+        await new Promise<void>((resolve) => {
+          this.once("reset", () => {
+            resolve();
+          });
+        });
+      }
+      this.count++;
+      return fetch(input, init);
+    };
+  }
+
+  public getCount(): number {
+    return this.count;
+  }
+}
+
+/**
+ * For the purposes of sync, organizers are isolated from each other.
+ * A failure for one organizer should not prevent other organizers from
+ * completing sync. Organizers also have independent rate limits with
+ * Pretix.
+ */
+export class OrganizerSync {
+  // This queue allows the sync system to limit concurrent DB access
+  public dbQueue: PQueue;
+  private organizer: DevconnectPretixOrganizerConfig;
   private pretixAPI: IDevconnectPretixAPI;
+  private fetcher: LimitedFetcher;
+  // A promise which resolves only when sync fully completes
+  private fetchPromise?: Promise<"complete">;
   private rollbarService: RollbarService | null;
-  private semaphoreService: SemaphoreService;
   private db: Pool;
-  private timeout: NodeJS.Timeout | undefined;
-  private _hasCompletedSyncSinceStarting: boolean;
-
-  public get hasCompletedSyncSinceStarting(): boolean {
-    return this._hasCompletedSyncSinceStarting;
-  }
+  private fetchedData?: {
+    data: EventData;
+    event: DevconnectPretixEventConfig;
+  }[];
 
   public constructor(
-    context: ApplicationContext,
+    dbQueue: PQueue,
+    organizer: DevconnectPretixOrganizerConfig,
+    fetchesPerCycle: number,
```

Comment:
how long is a cycle?

Reply:
One minute. (Sync runs once per minute, and Pretix limits requests-per-minute).